### PR TITLE
letsexchange - add plugin

### PR DESCRIPTION
**Pull request by xcodex2010**
letsexchange - add plugin

### CODE REVIEW

File: src/swap/letsexchange.js [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+      break;
+    case 'BSC':
+      networkCode = 'BEP20';
+      break;
+    default:
+      break;
+  }
+
+  return networkCode;
+}
```
Comment (thehobbit85) [inline src/swap/letsexchange.js:104]: Don't use a 'switch case' function, use the 'getCodesWithMainnetTranscription' instead.
Here is the link to the function: https://github.com/EdgeApp/edge-exchange-plugins/blob/80fac2c1ca153234f613241e4a5e7ef788c2a731/src/swap-helpers.js#L188
And this is an example usage: https://github.com/EdgeApp/edge-exchange-plugins/blob/80fac2c1ca153234f613241e4a5e7ef788c2a731/src/swap/changenow.js#L105

**Review by thehobbit85 — CHANGES_REQUESTED** [review]

File: src/swap/letsexchange.js [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    const response = await fetchCors(url, { method: 'POST', body, headers })
+    if (!response.ok) {
+      if (response.status === 422) {
+        throw new SwapCurrencyError(
+          swapInfo,
+          request.fromCurrencyCode,
+          request.toCurrencyCode
+        )
+      }
+      throw new Error(`letsexchange returned error code ${response.status}`)
```
Comment (paullinator) [inline src/swap/letsexchange.js:128]: Are there any cases where the user might get an "over limit" error or "region not supported" error? If so, need to catch these and throw the appropriate error.

File: src/swap/letsexchange.js [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+          withdrawal_extra_id: null,
+          affiliate_id: initOptions.apiKey,
+          promocode: promoCode != null ? promoCode : '',
+          type: 'edge',
+          float: true,
+          isEstimate: false
+        }
+      })
+      log('sendReply' + sendReply)
+      const quoteInfo: QuoteInfo = sendReply
```
Comment (paullinator) [inline src/swap/letsexchange.js:246]: This should use a `cleaner`. See https://github.com/EdgeApp/edge-exchange-plugins/blob/master/src/swap/sideshift.js#L377

**Review by paullinator — CHANGES_REQUESTED** [review]

File: src/swap/letsexchange.js (hunk 3 above)
Comment (xcodex2010) [reply src/swap/letsexchange.js:222]: Is there an example?

File: src/swap/letsexchange.js (hunk 1 above)
Comment (xcodex2010) [reply src/swap/letsexchange.js]: Fixed it.

File: src/swap/letsexchange.js (hunk 2 above)
Comment (xcodex2010) [reply src/swap/letsexchange.js:116]: No. There are not.

File: src/swap/letsexchange.js (hunk 3 above)
Comment (paullinator) [reply src/swap/letsexchange.js:222]: The link has an example

Comment (xcodex2010) [reply src/swap/letsexchange.js:222]: Fixed it.

File: src/swap/letsexchange.js [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+          promocode: promoCode != null ? promoCode : '',
+          type: 'edge',
+          float: true,
+          isEstimate: false
+        }
+      })
+
+      log('sendReply' + sendReply)
+      const quoteInfo = asQuoteInfo({
+        transaction_id: sendReply.transaction_id,
```
Comment (paullinator) [inline src/swap/letsexchange.js:229]: Not the right way to use a cleaner. You should take the raw data coming from an api call and pass it through a cleaner to validate the response.

ie 

```
const sendReply = await call(uri + 'transaction', request
const quoteInfo = asQuoteInfo(sendReply)
```

**Review by paullinator — CHANGES_REQUESTED** [review]

File: src/swap/letsexchange.js [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+} from '../swap-helpers.js'
+
+const pluginId = 'letsexchange'
+const swapInfo: EdgeSwapInfo = {
+  pluginId,
+  displayName: 'LetsExchange',
+  supportEmail: 'support@letsexchange.io'
+}
+
+const orderUri = 'https://api.letsexchange.io/?exchangeId='
```
Comment (paullinator) [inline src/swap/letsexchange.js:39]: This URL does not bring the user to a valid order status page.

**Review by paullinator — CHANGES_REQUESTED** [review]

File: src/swap/letsexchange.js (hunk 5 above)
Comment (xcodex2010) [reply src/swap/letsexchange.js:39]: this url show exchange info page, like this
image

File: src/swap/letsexchange.js (hunk 4 above)
Comment (xcodex2010) [reply src/swap/letsexchange.js]: Thank you! Got it. 
Done.

File: src/swap/letsexchange.js (hunk 5 above)
Comment (paullinator) [reply src/swap/letsexchange.js:39]: That was not my experience in testing. The links go to an API doc page.

**Comment by paullinator** [comment]
App crashes in testing. 

yarn precommit found this issue causing the crash.
```Monitor logs will go to /private/tmp/flow/zSUserszSpaulzSgitzSedge-exchange-plugins.monitor_log
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/swap/letsexchange.js:29:3

Cannot import getCodesWithMainnetTranscription because there is no getCodesWithMainnetTranscription export in
../swap-helpers.js.

     26│   type InvalidCurrencyCodes,
     27│   checkInvalidCodes,
     28│   makeSwapPluginQuote, getCodes,
     29│   getCodesWithMainnetTranscription
     30│ } from '../swap-helpers.js'
     31│
     32│ const pluginId = 'letsexchange'



Found 1 error
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
paul@Macallan edge-exchange-plugins % 
```
Please thoroughly test the plugin before submitting the PR.

**Review by paullinator — CHANGES_REQUESTED** [review]
Fix crash and thoroughly test

**Comment by paullinator** [comment]
please do not merge master in to you feature branch. makes it difficult to find changes. Please rebase your commits on top of master instead. As well, squash all commits into one as this is a small PR.

**Comment by paullinator** [comment]
in testing i could not swap BTC to ETH. Got this error
Screen Shot 2022-03-25 at 18 57 38

**Comment by xcodex2010** [comment]
Hi! Tell me please. How could i testing it without mobile app ?
Is it possible ?

**Comment by xcodex2010** [comment]
perhaps this is possible with **test/demo.js** ?

**Comment by paullinator** [comment]
No. You need the mobile app. See readme

**Comment by xcodex2010** [comment]
Hi!
Is there the step by step instruction for run and test exchange app ?
I did all that describe at readme. Run app via xcode emulator.
on edge-react-gui ran yarn ios. that ran app but without any exchanges at Settings > Exchange Settings. And could i get fake money on my test wallet at my app?

**Comment by xcodex2010** [comment]
Hi!
where can I get the log data that I describe in edge-exchange-plugins using the log() method;
How to debug my (edge-exchange-plugins) plugin ?

**Comment by xcodex2010** [comment]
Hi!. All done! Has been successfully tested! 
image

**Comment by paullinator** [comment]
Please do not create any merge commits into your feature branch. Please fix the PR and squash all commits into one with no merge commit. Instead of creating a merge from master to your branch, rebase your branch on top of master

**Review by paullinator — CHANGES_REQUESTED** [review]
see comment

**Comment by xcodex2010** [comment]
Sorry. I created new PR. Because this PR is like a trash already.

https://github.com/EdgeApp/edge-exchange-plugins/pull/183

**Comment by xcodex2010** [comment]
image

Does it matter? Or can we send it to you in another way ?

**Pull request closed without merging**